cmd/restic/cmd_copy.go

```diff
@@ -0,0 +1,214 @@
+
+	for sn := range FindFilteredSnapshots(ctx, dstRepo, opts.Hosts, opts.Tags, opts.Paths, nil) {
+		if sn.Original != nil && !sn.Original.IsNull() {
+			copiedSnapshots[*sn.Original] = sn.ID()
+		}
+	}
+
+	for sn := range FindFilteredSnapshots(ctx, srcRepo, opts.Hosts, opts.Tags, opts.Paths, args) {
+		Verbosef("\nsnapshot %s of %v at %s)\n", sn.ID().Str(), sn.Paths, sn.Time)
+
+		if newID, ok := copiedSnapshots[*sn.ID()]; ok {
+			Verbosef("skipping source snapshot %s, was already copied to snapshot %s\n", sn.ID().Str(), newID.Str())
+			continue
+		}
+		if sn.Original != nil && dstRepo.Has(restic.Handle{Type: restic.SnapshotFile, Name: sn.Original.String()}) {
+			Verbosef("skipping source snapshot %s, which is a copy of snapshot %s\n", sn.ID().Str(), sn.Original.Str())
+			continue
+		}
+		Verbosef("  copy started, this may take a while...\n")
+
+		if err := copySnapshot(ctx, srcRepo, dstRepo, *sn.Tree, visitedTrees); err != nil {
+			return err
+		}
+		debug.Log("tree copied")
+
+		if err = dstRepo.Flush(ctx); err != nil {
+			return err
+		}
+		debug.Log("flushed packs and saved index")
+
+		// save snapshot
+		sn.Parent = nil // Parent does not have relevance in the new repo.
+		// Original does not have relevance in the new repo. Remember the source snapshot instead
+		sn.Original = sn.ID()
+		newID, err := dstRepo.SaveJSONUnpacked(ctx, restic.SnapshotFile, sn)
+		if err != nil {
+			return err
+		}
+		Verbosef("snapshot %s saved\n", newID.Str())
+	}
+	return nil
+}
+
+func copySnapshot(ctx context.Context, srcRepo, dstRepo restic.Repository, treeID restic.ID, visitedTrees restic.IDSet) error {
+	// We have already processed this tree
+	if visitedTrees.Has(treeID) {
+		return nil
+	}
+
+	tree, err := srcRepo.LoadTree(ctx, treeID)
+	if err != nil {
+		return fmt.Errorf("LoadTree(%v) returned error %v", treeID.Str(), err)
+	}
+	visitedTrees.Insert(treeID)
+
+	// Do we already have this tree blob?
+	if !dstRepo.Index().Has(treeID, restic.TreeBlob) {
+		newTreeID, err := dstRepo.SaveTree(ctx, tree)
+		if err != nil {
+			return fmt.Errorf("SaveTree(%v) returned error %v", treeID.Str(), err)
+		}
+		// Assurance only.
+		if newTreeID != treeID {
+			return fmt.Errorf("SaveTree(%v) returned unexpected id %s", treeID.Str(), newTreeID.Str())
+		}
+	}
+
+	// TODO: keep only one (big) buffer around.
+	// TODO: parellize this stuff, likely only needed inside a tree.
+
+	for _, entry := range tree.Nodes {
+		// If it is a directory, recurse
+		if entry.Type == "dir" && entry.Subtree != nil {
+			if err := copySnapshot(ctx, srcRepo, dstRepo, *entry.Subtree, visitedTrees); err != nil {
+				return err
+			}
+		}
+		// Copy the blobs for this file.
+		for _, blobID := range entry.Content {
+			// Do we already have this data blob?
+			if dstRepo.Index().Has(blobID, restic.DataBlob) {
+				continue
+			}
+			debug.Log("Copying blob %s\n", blobID.Str())
+			buf, err := srcRepo.LoadBlob(ctx, restic.DataBlob, blobID, nil)
```

Comment:
The complexity here is that we have to (well, should) reuse a single `buf` across all `copyTree` recursions. Otherwise each recursion level could hold on to a buffer with a few megabytes each. One rather ugly option would be to pass a `buf *[]byte` parameter along the recursion. Do you have a better idea?

Reply:
For this implementation (with parallelizing it already on the aganda) I would just use a global `var buf []bytes` - maybe with a comment `// TODO: change this when parallelizing`.

Reply:
Ugh, I really dislike global variables. I've decided to move the repository objects and visitedTrees into a struct.